sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -415,16 +422,34 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     }
 
     if (DDLUtils.isDatasourceTable(withStatsProps)) {
-      val oldDef = client.getTable(db, withStatsProps.identifier.table)
-      // Sets the `schema`, `partitionColumnNames` and `bucketSpec` from the old table definition,
-      // to retain the spark specific format if it is. Also add old data source properties to table
-      // properties, to retain the data source table format.
-      val oldDataSourceProps = oldDef.properties.filter(_._1.startsWith(DATASOURCE_PREFIX))
+      val oldTableDef = client.getTable(db, withStatsProps.identifier.table)
+
+      val oldLocation = getLocationFromRawTable(oldTableDef)
+      // Only update the `locationUri` field if the location is really changed, because this table
+      // may be not Hive-compatible and can not set the `locationUri` field. We should respect the
+      // old `locationUri` even it's None.
+      val newLocation = if (oldLocation == tableDefinition.storage.locationUri) {
+        oldTableDef.storage.locationUri
+      } else {
+        tableDefinition.storage.locationUri
+      }
```

Comment:
Just FYI, we do not have any test case to cover this. `ALTER TABLE SET LOCATION` for data source tables is only tested in `DDLSuite` (when Hive support is not enabled). That means, these code changes are not tested. Actually, this is not the only one. : (